tests/model_selection/test_hyperband.py

```diff
@@ -0,0 +1,431 @@
+    max_iter = 27
+
+    alg = HyperbandSearchCV(
+        model, params, max_iter=max_iter, patience=True, random_state=0
+    )
+    yield alg.fit(X, y)
+
+    alg_patience = max_iter // alg.aggressiveness
+    actual_decisions = [b.pop("decisions") for b in alg.metadata_["brackets"]]
+    paper_decisions = [b.pop("decisions") for b in alg.metadata["brackets"]]
+
+    for paper_iter, actual_iter in zip(paper_decisions, actual_decisions):
+        trimmed_paper_iter = {k for k in paper_iter if k <= alg_patience}
+        assert trimmed_paper_iter.issubset(set(actual_iter))
+        assert all(x <= alg_patience + 1 for x in actual_iter)
+
+    assert alg.metadata_["partial_fit_calls"] <= alg.metadata["partial_fit_calls"]
+    assert alg.best_score_ >= 0.9
+
+    alg = HyperbandSearchCV(model, params, max_iter=max_iter, patience=1)
+    with pytest.warns(UserWarning, match="The goal of `patience`"):
+        yield alg.fit(X, y)
+
+
+@gen_cluster(client=True, timeout=5000)
+def test_cv_results_order_preserved(c, s, a, b):
+    X, y = make_classification(n_samples=10, n_features=4, chunks=10)
+    model = ConstantFunction()
+    params = {"value": scipy.stats.uniform(0, 1)}
+    alg = HyperbandSearchCV(model, params, max_iter=9, random_state=42)
+    yield alg.fit(X, y)
+
+    info = {k: v[-1] for k, v in alg.model_history_.items()}
+    for _, row in pd.DataFrame(alg.cv_results_).iterrows():
+        model_info = info[row["model_id"]]
+        assert row["bracket"] == model_info["bracket"]
+        assert row["params"] == model_info["params"]
+        assert np.allclose(row["test_score"], model_info["score"])
+
+
+@gen_cluster(client=True, timeout=5000)
+def test_integration(c, s, a, b):
+    X, y = make_classification(n_samples=10, n_features=4, chunks=10)
+    model = ConstantFunction()
+    params = {"value": scipy.stats.uniform(0, 1)}
+    alg = HyperbandSearchCV(model, params, max_iter=9, random_state=42)
+    yield alg.fit(X, y)
+    cv_res_keys = set(alg.cv_results_.keys())
+    gt_zero = lambda x: x >= 0
+    gt_one = lambda x: x >= 1
+    for column, dtype, condition in [
+        ("params", dict, lambda d: set(d.keys()) == {"value"}),
+        ("test_score", float, gt_zero),
+        ("rank_test_score", int, gt_one),
+        ("mean_partial_fit_time", float, gt_zero),
+        ("std_partial_fit_time", float, gt_zero),
+        ("mean_score_time", float, gt_zero),
+        ("std_score_time", float, gt_zero),
+        ("model_id", None, lambda x: isinstance(x, str)),
+        ("partial_fit_calls", int, gt_zero),
+        ("param_value", float, gt_zero),
+        ("bracket", int, None),
+    ]:
+        if dtype:
+            assert alg.cv_results_[column].dtype == dtype
+        if condition:
+            assert all(condition(x) for x in alg.cv_results_[column])
+        cv_res_keys -= {column}
+
+    # the keys listed in the for-loop are all the keys in cv_results_
+    assert cv_res_keys == set()
```

Comment:
Expand this to make it more transparent.

``` python
columns = [...]
for column, dtype, condition in columns:
    ...
assert columns.keys() == alg.cv_results_.keys()
```

Reply:
In the docstring, expand why this test is needed: Hyperband is a wrapper around a couple different IncrementalSearchCV objects.